### ISSUE TITLE
Invalidate cache when rules change

### ISSUE THREAD

**Issue by TheSavior** [eslint/eslint#3770]
We should automatically invalidate the cache when the configuration used to test the file was changed. This will allow people to use the cache on continuous integration machines that do incremental builds (opposed to travis which does a full checkout each time). 

In order to support this I think we would need to store the full configuration each file was linted with in the eslintcache file. That would provide support for upstream plugins being changed, environments in higher folders being changed, etc.

**Comment by eslintbot** [comment]
Thanks for the issue! If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

**Comment by lo1tuma** [comment]
:+1: 

I’ve already mentioned this in our chat:

> When using the --cache option I wonder if the cache gets invalidated if I change my eslint config (.eslintrc, package.json or CLI arguments) or if I update/downgrade ESLint.

-----

A possible solution for this could be to store a checksum of `eslintVersion + getEslintConfigForFile(file) + fileContent`, each time the checksum for a file is different the cache should be invalidated.

**Comment by TheSavior** [comment]
Yeah, that approach sounds right to me.

**Comment by nzakas** [comment]
@royriojas what do you think?

**Comment by royriojas** [comment]
I agree. I don't think we need to store the configuration per file since the cache is created per run. The checksum of the configuration + the version of eslint might be enough. 

My only question is if plugin versions are also available to the cliEngine. I can take a look to this tonight if you want, it seems a pretty reasonable request

**Comment by lo1tuma** [comment]
> I don't think we need to store the configuration per file since the cache is created per run.

I don't understand what you mean with this. If the cache is created from scratch for each run, why would we need caching at all.?

**Comment by royriojas** [comment]
@lo1tuma the cache is created per run only if it does not exists already. If it exists then we honor it. 

What I really tried to say is that the configuration is the same for all the files, at least that was what I saw during the testing for the implementation. Maybe @nzakas can chime in and provide details if there is any case that the configuration might change per file.

An easy solution is to add the eslint-version + hashOfconfig object, if this value changes then we ignore the cache.

I guess that will work fine at least for 80% of the cases. A corner case will be if a plugin changes its behavior between versions, in that case the cache will not affect those files. @nzakas is there a way to know the version of the plugins being used? should I try to figure it out the version from the package.json files?

**Comment by lo1tuma** [comment]
The config can be different for each file, as you can have multiple `.eslintrc` files, one in each folder. Have a look to the documentation how the [configuration cascading an hierarchy](http://eslint.org/docs/user-guide/configuring#configuration-cascading-and-hierarchy) works. We also plan to add glob-pattern based configuration support. So you should always assume that the configuration can be different for each file.

I agree that we should only store a checksum and not the whole configuration:

```
checksum(eslintVersion + getConfigForFile(file) + fileContent);
```

If we are going to use a checksum anyway we could also include the file content and also get rid of the `mtime` based cache.

-----

Getting the versions of the used plugins is not so easy but it should be possible, something like:

```js
var config = getConfigForFile(file);
var plugins = config.plugins;
var path = require('path');

var pluginVersion = plugins.map(function (pluginName) {
    var pathToPlugin = path.dirname(require.resolve(pluginName)); // make sure pluginName is prefixed with 'eslint-plugin-'
    var packageConfig = require(path.join(pathToPlugin, 'package.json'));
    return packageConfig.version;
}); 
```

**Comment by royriojas** [comment]
> If we are going to use a checksum anyway we could also include the file content and also get rid of the mtime based cache.

Only concern with adding the `fileContent` has is the increase of time to calculate it.

If that's not an issue I can add that with ease as well

**Comment by nzakas** [comment]
The whole point of adding caching is to speed things up, if we it ends up slower because we have to calculate hashes for every file, then we need to measure the difference and determine if it's worth it. Performance is the feature here, so we can't just throw that out to trade off for correctness (in that case, we'd just remove the feature if it can't be both fast and correct).

I don't think we can ever rely on plugin versions for anything meaningful, especially since we'll likely have a few different ways of installing plugins in the future.

**Comment by TheSavior** [comment]
What is wrong with a hash of the configuration and the file mtime?

**Comment by royriojas** [comment]
@theSavior IMHO, nothing is wrong with the file `mtime`/`size` fields. The only concern is that CVS tools like git sometimes update the `mtime` during a checkout. I don't see that as really big concern, because worst case scenario only a few files will be considered not cached, which is I guess safe, they will be just cached the next time. 

I was planning to add the hash of the configuration per file, that should not be that costly and will have the better effects I guess.

**Comment by lo1tuma** [comment]
Ok, so using `mtime` is not really related to this issue, I will probably file a separate issue for that where we can evaluate how expensive it would be to create a checksum for the file content.

**Comment by royriojas** [comment]
I made a first draft of this feature here: https://github.com/eslint/eslint/pull/3818.
